modules/core/src/main/java/org/apache/ignite/internal/processors/cache/ExchangeContext.java

```diff
@@ -58,7 +66,17 @@
     public ExchangeContext(boolean crd, GridDhtPartitionsExchangeFuture fut) {
         int protocolVer = exchangeProtocolVersion(fut.firstEventCache().minimumNodeVersion());
 
-        if (compatibilityNode || (crd && fut.localJoinExchange())) {
+        GridDiscoveryManager disco = fut.sharedContext().discovery();
+
+        if (protocolVer > 2 && fut.exchangeId().isLeft() && fut.isFirstEventNodeInBaseline() &&
```

Comment:
Protocol v3 should be disabled if previous assignment is not ideal. Otherwise it means rebalancing is not finished and some nodes can see different local partition states at the moment of assignment calculation.